### PR TITLE
Share the list of gem dependencies between Linux and Windows

### PR DESCRIPTION
**Pull request by jkohen**
This change bumps the version of fluentd and of some gems on the Windows side, as now we ship the same in both platforms.

I verified that the installer gives a working agent: I installed the agent and immediately saw logs in the Stackdriver UI.

### CODE REVIEW

File: bin/gem_installer [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env ruby
```
Comment (igorpeshansky) [inline bin/gem_installer]: This script is basically a copy of `bin/gem_downloader`... The only differences I can see are:
- this doesn't support `:fetch`,
- this doesn't use `--explain` in the `gem install` command.

Can we please add some comments on why we need both scripts?

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 2]
```diff
@@ -135,18 +144,15 @@ $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";"
...
-# We install all gems with '--no-document' so we don't pull in
-# unneeded docs that bloat the file size (and also seem to cause issues with unzipping).
 ###############################
 
-& $GEM_CMD install fluentd:1.4.2 --no-document
-& $GEM_CMD install windows-pr:1.2.6 win32-ipc:0.7.0 win32-event:0.6.3 win32-eventlog:0.6.7 win32-service:2.1.4 fluent-plugin-windows-eventlog:0.2.2 --no-document
-& $GEM_CMD install google-protobuf:3.7.1 grpc:1.20.0 fluent-plugin-google-cloud:0.7.23 --no-document
+$gem_installer = $SRC_ROOT + '\bin\gem_installer'
+$plugin_gems_rb = $SRC_ROOT + '\plugin_gems.rb'
+& $RUBY_EXE $gem_installer $plugin_gems_rb
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: Both `grpc` and `google-protobuf` are in `core_gems.rb`. Don't we need to pull `core_gems.rb` as well?

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 3]
```diff
@@ -135,18 +144,15 @@ $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";"
...
+# Install the needed gems for google fluentd to works.
 #
 # These are all a known set of versions which work together. Be sure things work if you
 # update any of them.
-#
-# We install all gems with '--no-document' so we don't pull in
-# unneeded docs that bloat the file size (and also seem to cause issues with unzipping).
 ###############################
 
-& $GEM_CMD install fluentd:1.4.2 --no-document
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: The Linux build does not install `fluentd` via the `gem` command, but rather has omnibus build it from source (see [config](https://github.com/GoogleCloudPlatform/google-fluentd/blob/master/config/software/fluentd.rb)). Since we don't use omnibus on Windows, should we not add it to `plugin_gems.rb` windows section? If not, where do you install the `fluentd` gem?

File: plugin_gems.rb [hunk 4]
```diff
@@ -27,3 +27,11 @@
 download "fluent-plugin-systemd", "1.0.2"
 download "opencensus", "0.5.0"
 download "opencensus-stackdriver", "0.3.0"
+if windows?
+  download 'windows-pr', '1.2.6'
```
Comment (igorpeshansky) [inline plugin_gems.rb]: Are single quotes significant? If not, can they be consistent with the rest of the file?

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 5]
```diff
@@ -135,18 +144,15 @@ $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";"
 #  STEP 4 - INSTALL THE GEMS
 ##############################
 #
-# Install the needed gems for google fleuntd to works.
+# Install the needed gems for google fluentd to works.
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: Since you're fixing this anyway, might as well `s/works/work/`...

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 6]
```diff
@@ -199,3 +205,5 @@ cp $NSIS_UNZU_DLL $NSIS_UNICODE_PLUGIN_DIR
 ##############################
 
 & $NSIS_MAKE /DVERSION=$version $STACKDRIVER_NSI
+
+Set-PSDebug -Trace 0
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: Just curious, is this a cleanup step that was missing before?

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 7]
```diff
@@ -109,10 +111,17 @@ mkdir $NSIS_UNZU_DIR
 $ProgressPreference = "silentlyContinue"
 # Handle SSL correctly.
 [Net.ServicePointManager]::SecurityProtocol = 'TLS12'
-# Pretend to be curl for Sourceforge redirects to work.
-Invoke-WebRequest "$RUBY_INSTALLER_LINK" -OutFile "$RUBY_INSTALLER" -UserAgent "curl/7.60.0"
-Invoke-WebRequest "$NSIS_INSTALLER_LINK" -OutFile "$NSIS_INSTALLER" -UserAgent "curl/7.60.0"
-Invoke-WebRequest "$NSIS_UNZU_INSTALLER_LINK" -OutFile "$NSIS_UNZU_ZIP" -UserAgent "curl/7.60.0"
+
+# Need help with Chocolatey? See https://chocolatey.org/install
```
[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: What makes you think it's incomplete?
The reason is in the comment: if anyone needs help because that step fails, they should look at that page. I don't have the ability to test all the possible environments where this could run.

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 8]
```diff
@@ -109,10 +111,17 @@ mkdir $NSIS_UNZU_DIR
...
+Set-ExecutionPolicy Bypass -Scope Process -Force; iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
```
[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: Yes, we can.

[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: Can you explain why it should be `AllSigned`? I took the recipe straight from the Chocolatey "how to install Chocolatey" page, and it works. That passed my bar for correctness. Did I miss something?

[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: I took the recipe straight from the Chocolatey "how to install Chocolatey" page, and it worked. I've splitted it.

[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: If Chocolatey is already installed, it just prints a notice about that fact. Is that an issue?

I don't understand the way the steps are divided. If you think it's important, please explain the criteria that we should use for that.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 3 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: I can add it, if you think it's important. This is the list of gems that are getting installed as it is now, including fluentd. It probably gets installed as a dependency for something else:
addressable (2.7.0)
aws-eventstream (1.0.3)
aws-partitions (1.149.0)
aws-sdk-core (3.48.3)
aws-sdk-kms (1.16.0)
aws-sdk-s3 (1.36.0)
aws-sdk-sqs (1.13.0)
aws-sigv4 (1.1.0)
bigdecimal (default: 1.3.4)
certstore_c (0.1.6)
cmath (default: 1.0.0)
concurrent-ruby (1.1.5)
cool.io (1.5.4 x86-mingw32)
csv (default: 1.0.0)
date (default: 1.0.0)
dbm (default: 1.0.0)
declarative (0.0.10)
declarative-option (0.1.0)
did_you_mean (1.2.0)
dig_rb (1.0.1)
domain_name (0.5.20190701)
etc (default: 1.0.0)
faraday (0.17.0)
fcntl (default: 1.0.0)
ffi (1.11.1 x86-mingw32)
ffi-compiler (1.0.1)
ffi-win32-extensions (1.0.3)
fiddle (default: 1.0.0)
fileutils (default: 1.0.2)
fluent-config-regexp-type (1.0.0)
fluent-plugin-detect-exceptions (0.0.12)
fluent-plugin-google-cloud (0.7.23)
fluent-plugin-kubernetes_metadata_filter (2.4.0)
fluent-plugin-multi-format-parser (1.0.0)
fluent-plugin-prometheus (1.4.0)
fluent-plugin-record-modifier (2.0.1)
fluent-plugin-record-reformer (0.9.1)
fluent-plugin-rewrite-tag-filter (2.2.0)
fluent-plugin-s3 (1.1.10)
fluent-plugin-systemd (1.0.2)
fluent-plugin-webhdfs (1.2.3)
fluentd (1.7.4 x86-mingw32, 1.6.3 x86-mingw32)
gdbm (default: 2.0.0)
google-api-client (0.30.8)
google-cloud-core (1.4.1)
google-cloud-env (1.3.0)
google-cloud-logging (1.6.6)
google-cloud-monitoring (0.29.5)
google-cloud-trace (0.36.0)
google-gax (1.8.1, 1.7.1)
google-protobuf (3.9.0 x86-mingw32)
googleapis-common-protos (1.3.9)
googleapis-common-protos-types (1.0.4)
googleauth (0.9.0)
grpc (1.24.0 x86-mingw32, 1.22.0 x86-mingw32)
http (4.2.0)
http-accept (1.7.0)
http-cookie (1.0.3)
http-form_data (2.1.1)
http-parser (1.2.1)
http_parser.rb (0.6.0)
httpclient (2.8.3)
io-console (default: 0.4.6)
ipaddr (default: 1.2.0)
jmespath (1.4.0)
json (2.2.0, default: 2.1.0)
jwt (2.2.1)
kubeclient (4.5.0)
lru_redux (1.1.0)
memoist (0.16.0)
mime-types (3.3)
mime-types-data (3.2019.1009)
mini_mime (1.0.2)
minitest (5.10.3)
msgpack (1.3.1 x86-mingw32)
multi_json (1.14.1)
multipart-post (2.1.1)
net-telnet (0.1.1)
netrc (0.11.0)
opencensus (0.5.0)
opencensus-stackdriver (0.3.0)
openssl (default: 2.1.2)
os (1.0.1)
power_assert (1.1.1)
prometheus-client (0.9.0)
psych (default: 3.0.2)
public_suffix (4.0.1)
quantile (0.2.1)
rake (12.3.0)
rdoc (default: 6.0.1)
recursive-open-struct (1.1.0)
representable (3.0.4)
rest-client (2.1.0 x86-mingw32)
retriable (3.1.2)
rly (0.2.3)
scanf (default: 1.0.0)
sdbm (default: 1.0.0)
serverengine (2.1.1 x86-mingw32)
sigdump (0.2.4)
signet (0.12.0)
stackdriver-core (1.4.0)
stringio (default: 0.0.1)
strptime (0.2.3)
strscan (default: 1.0.0)
systemd-journal (1.3.3)
test-unit (3.2.7)
thread_safe (0.3.6)
tzinfo (2.0.0, 1.2.5)
tzinfo-data (1.2019.3)
uber (0.1.0)
unf (0.1.4)
unf_ext (0.0.7.6)
uuidtools (2.1.5)
webhdfs (0.8.0)
webrick (default: 1.4.2)
win32-api (1.8.0 universal-mingw32)
win32-event (0.6.3)
win32-ipc (0.6.6)
win32-service (0.8.10)
windows-api (0.4.4)
windows-pr (1.2.6)
xmlrpc (0.3.0)
yajl-ruby (1.4.1)
zlib (default: 1.0.0)

File: plugin_gems.rb (hunk 4 above)
Comment (jkohen) [reply plugin_gems.rb]: It works.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 6 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: This step wasn't there. If you're asking what happens without this step, then the powershell session that spawned the script inherits the trace flag, which makes it really annoying to use it interactively.

As an aside, I find that having the Trace output makes it really hard for me to see errors, which has caused me to miss problems while I was developing the script. I think it's a bad practice to leave these on in most submitted scripts.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 2 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: Those dependencies are already being pulled in. See the list in the other comment below.

File: bin/gem_installer (hunk 1 above)
Comment (jkohen) [reply bin/gem_installer]: Because packaging on Linux and Windows still works differently. The Linux one doesn't seem to install the gems, hence the `--explain`. I can add a `--install` flag to the downloader and the feature, is that what you're asking me to do?

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 6 above)
Comment (igorpeshansky) [reply windows-installer/generate_sdl_agent_exe.ps1]: Would it be possible to add a comment that this is cleanup for interactive usage?

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 9]
```diff
@@ -177,6 +188,7 @@ rm -r -Force $RUBY_DEV_DIR
 ##############################
 
 Add-Type -Assembly System.IO.Compression.FileSystem
+rm -Force $STACKDRIVER_ZIP -ErrorAction Ignore
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: Can you please move this `rm` above the `Add-Type` (which is setup for the next line)?
Also, using the `rm` alias with the Windows option syntax (`-ErrorAction`) looks weird — can it be `Remove-Item` instead?

File: bin/gem_installer (hunk 1 above)
Comment (jkohen) [reply bin/gem_installer]: The script is now different, without dependencies on curl, and stands on its own. Note that I copied the script from omnibus-tb and preserved the naming there, so I don't see a reason to change it.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: Removed code.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 6 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: I'm confused, why is the comment needed here and not where Trace is set to 1? How is the comment improving readability of the script?

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 9 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: 1. done
2. the line above uses `rm -Force`, so I was being consistent, following your previous feedback.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: Removed this block.

File: core_gems.rb [hunk 10]
```diff
@@ -1,13 +1,16 @@
...
 download "sigdump", "0.2.4"
 download "thread_safe", "0.3.5"
 download "oj", "3.3.10"
 download "tzinfo", "1.2.2"
 download "tzinfo-data", "1.2016.5"
-fetch "google-protobuf", "3.7.1"
-fetch "grpc", "1.20.0"
+unless windows?
+  # Build dependencies for 'google-fluentd' that aren't needed on Windows.
+  fetch "google-protobuf", "3.9.0"
```
Comment (qingling128) [inline core_gems.rb]: [Optional] It might be helpful to delay gem version bumps unless they are required by this change. Because they might introduce extra noise and potentially break nightly jobs (grpc and protobuf updates had broken us multiple times in the past). This way it's clear if a certain issue is isolated to this refactoring only or some new gem dependencies.

File: core_gems.rb [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
 download "yajl-ruby", "1.3.1"
 download "sigdump", "0.2.4"
 download "thread_safe", "0.3.5"
 download "oj", "3.3.10"
 download "tzinfo", "1.2.2"
 download "tzinfo-data", "1.2016.5"
-fetch "google-protobuf", "3.7.1"
-fetch "grpc", "1.20.0"
+unless windows?
+  # Build dependencies for 'google-fluentd' that aren't needed on Windows.
```
Comment (igorpeshansky) [inline core_gems.rb]: More precisely, they are needed on Windows, but are installed through another mechanism.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 6 above)
Comment (igorpeshansky) [reply windows-installer/generate_sdl_agent_exe.ps1]: It wasn't very clear why changing some value that is intended to be script-scope as the last step in a script has any effect... But this is not a big deal either way; please treat as optional.

File: windows-installer/generate_sdl_agent_exe.ps1 [hunk 12]
```diff
@@ -135,18 +139,18 @@ $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";"
...
-# unneeded docs that bloat the file size (and also seem to cause issues with unzipping).
 ###############################
 
+$gem_installer = $SRC_ROOT + '\bin\gem_installer'
+$core_gems_rb = $SRC_ROOT + '\core_gems.rb'
+$plugin_gems_rb = $SRC_ROOT + '\plugin_gems.rb'
 & $GEM_CMD install fluentd:1.4.2 --no-document
-& $GEM_CMD install windows-pr:1.2.6 win32-ipc:0.7.0 win32-event:0.6.3 win32-eventlog:0.6.7 win32-service:2.1.4 fluent-plugin-windows-eventlog:0.2.2 --no-document
-& $GEM_CMD install google-protobuf:3.7.1 grpc:1.20.0 fluent-plugin-google-cloud:0.7.23 --no-document
+& $RUBY_EXE $gem_installer $core_gems_rb
```
Comment (igorpeshansky) [inline windows-installer/generate_sdl_agent_exe.ps1]: Now that there are no Windows-specific core_gems changes, we probably don't need this anymore.

File: bin/gem_installer (hunk 1 above)
Comment (igorpeshansky) [reply bin/gem_installer]: The script is still 90% similar.

That said, what's "omnibus-tb"? Do you mean treasure-data/omnibus-td-agent? If so, would it not make sense to copy a (modified) script from our fork, for consistency, rather than a (potentially diverged) script from the upstream repo?

In any case, I will send a follow-up PR to unify the scripts, so we can leave it as-is in this PR.

File: core_gems.rb [hunk 13]
```diff
@@ -1,13 +1,16 @@
 dir 'core_gems'
 download "json", "2.1.0"
-download "msgpack", "1.2.2"
-download "cool.io", "1.5.3"
+download "msgpack", "1.2.9"
```
Comment (igorpeshansky) [inline core_gems.rb]: Are these version bumps required for building on Windows? If not, can we please pull them out into a separate PR?

File: core_gems.rb (hunk 11 above)
Comment (jkohen) [reply core_gems.rb]: Thanks for the clarification.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 12 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: I'll leave it this way for consistency. We'll build something closer to what we build on Linux, maybe... and if anything breaks specifically on Windows, we'll know.

File: core_gems.rb (hunk 13 above)
Comment (jkohen) [reply core_gems.rb]: Yes, they are required.

File: bin/gem_installer (hunk 1 above)
Comment (jkohen) [reply bin/gem_installer]: Yes, I meant `treasure-data/omnibus-td-agent`, I'm sorry for confusing you. Thanks for all the ideas, I think this is fine.

File: windows-installer/generate_sdl_agent_exe.ps1 (hunk 6 above)
Comment (jkohen) [reply windows-installer/generate_sdl_agent_exe.ps1]: It's not clear to me either, and if the script fails the command doesn't run. But when it runs, it's useful :) First thing I do when I work with this script now is to comment out the first Trace command, so I can actually see what the script is doing. YMMV.

File: core_gems.rb (hunk 11 above)
Comment (igorpeshansky) [reply core_gems.rb]: I was proposing a change for a misleading comment.

Comment (igorpeshansky) [reply core_gems.rb]: I was proposing a change for a misleading comment. How about:
```
  # Build dependencies for 'google-fluentd' that don't need to be fetched explicitly on Windows.
```
?
Or, since this whole file is build dependencies for google-fluentd:
```
  # Gems that don't need to be fetched explicitly on Windows.
```
?

**Comment by jkohen** [comment]
I applied your suggestion @igorpeshansky . Can you please approve this PR?
